### ISSUE TITLE
Improve AWS Config implementation to reduce the number of AWS API calls

### ISSUE THREAD

**Issue by CarlosRS9** [wazuh/wazuh#10845]
|Component|
|---|
| AWS module |

## Description

The `AWS Config` integration perform multiple requests to the AWS API to obtain the list of logs to process. In particular, for AWS Config, the reason is because the `build_s3_filter_args` function is being invoked several times, each one using a different `date` value. This is the only AWS bucket implementation doing that. This function is invoked as many times as dates available to process.

We should improve the design of this implementation to reduce the number of calls to `build_s3_filter_args` and `list_objects_v2`, as it adds an overhead in the module performance.

Here is an example of execution. Every time the message `Marker: AWSLogs/XXXXXXXXXXXXXXXXXXXXXXXX` is present it means a new call to `build_s3_filter_args` and `list_objects_v2` has been made:
```

root@da140fd275b3:/var/ossec/wodles/aws# ./aws-s3 -b wazuh-aws-wodle-config -d2 -t config  -s 2021-oct-17 
DEBUG: +++ Debug mode on - Level: 2
DEBUG: +++ Working on XXXXXXXXXXXX - us-east-1
DEBUG: +++ Marker: AWSLogs/XXXXXXXXXXXX/Config/us-east-1/2021/10/19/ConfigHistory/XXXXXXXXXXXX_Config_us-east-1_ConfigHistory_AWS::Config::ResourceCompliance_20200601T004303Z_20211019T025123Z_1.json.gz
DEBUG: ++ Skipping previously processed file: AWSLogs/XXXXXXXXXXXX/Config/us-east-1/2021/10/19/ConfigSnapshot/XXXXXXXXXXXX_Config_us-east-1_ConfigSnapshot_20200601T105550Z_6e2fd7f4-e9dd-4730-b0da-d1f90ab2b363.json.gz
DEBUG: +++ DB Maintenance
DEBUG: +++ Marker: AWSLogs/XXXXXXXXXXXX/Config/us-east-1/2021/10/20
DEBUG: +++ No logs to process in bucket: XXXXXXXXXXXX/us-east-1
DEBUG: +++ Marker: AWSLogs/XXXXXXXXXXXX/Config/us-east-1/2021/10/21
DEBUG: +++ No logs to process in bucket: XXXXXXXXXXXX/us-east-1
DEBUG: +++ Marker: AWSLogs/XXXXXXXXXXXX/Config/us-east-1/2021/10/22
DEBUG: +++ No logs to process in bucket: XXXXXXXXXXXX/us-east-1
DEBUG: +++ Marker: AWSLogs/XXXXXXXXXXXX/Config/us-east-1/2021/10/23
DEBUG: +++ No logs to process in bucket: XXXXXXXXXXXX/us-east-1
DEBUG: +++ Marker: AWSLogs/XXXXXXXXXXXX/Config/us-east-1/2021/10/24
DEBUG: +++ No logs to process in bucket: XXXXXXXXXXXX/us-east-1
DEBUG: +++ Marker: AWSLogs/XXXXXXXXXXXX/Config/us-east-1/2021/10/25
DEBUG: +++ No logs to process in bucket: XXXXXXXXXXXX/us-east-1
DEBUG: +++ Marker: AWSLogs/XXXXXXXXXXXX/Config/us-east-1/2021/10/26
DEBUG: +++ No logs to process in bucket: XXXXXXXXXXXX/us-east-1
DEBUG: +++ Marker: AWSLogs/XXXXXXXXXXXX/Config/us-east-1/2021/10/27
DEBUG: +++ No logs to process in bucket: XXXXXXXXXXXX/us-east-1
DEBUG: +++ Marker: AWSLogs/XXXXXXXXXXXX/Config/us-east-1/2021/10/28
DEBUG: +++ No logs to process in bucket: XXXXXXXXXXXX/us-east-1
DEBUG: +++ Marker: AWSLogs/XXXXXXXXXXXX/Config/us-east-1/2021/10/29
DEBUG: +++ No logs to process in bucket: XXXXXXXXXXXX/us-east-1
DEBUG: +++ Marker: AWSLogs/XXXXXXXXXXXX/Config/us-east-1/2021/10/30
DEBUG: +++ No logs to process in bucket: XXXXXXXXXXXX/us-east-1
DEBUG: +++ Marker: AWSLogs/XXXXXXXXXXXX/Config/us-east-1/2021/10/31
DEBUG: +++ No logs to process in bucket: XXXXXXXXXXXX/us-east-1
DEBUG: +++ Marker: AWSLogs/XXXXXXXXXXXX/Config/us-east-1/2021/11/1
DEBUG: +++ No logs to process in bucket: XXXXXXXXXXXX/us-east-1
DEBUG: +++ Marker: AWSLogs/XXXXXXXXXXXX/Config/us-east-1/2021/11/2
DEBUG: +++ No logs to process in bucket: XXXXXXXXXXXX/us-east-1
DEBUG: +++ Marker: AWSLogs/XXXXXXXXXXXX/Config/us-east-1/2021/11/3
DEBUG: +++ No logs to process in bucket: XXXXXXXXXXXX/us-east-1
DEBUG: +++ DB Maintenance
```

**Comment by noise-kngdm** [comment]
## Issue update
Right now, the module is making a great amount of `list_objects_v2` calls per execution because the folder structure that the AWS Config module uses to store the events in a bucket doesn't use the 0 padded date notation. That means that it will store an event corresponding to July 7th using a key like the following:  `wazuh-aws-wodle-configAWSLogs/XXXXXXXXXXXX/Config/us-east-1/2021/7/7/07/ConfigHistory/166157441623_Config_us-east-1_ConfigHistory_AWS::Config::ResourceCompliance_20210707T004303Z_20210707T025123Z_1.json.gz`.  

Because of that, the log files can't be filtered using the `StartAfter` parameter, since the AWS API uses this parameter to compare strings and we are interested in start fetching after a given date. With the AWS API's implementation, a log corresponding to November 11th would be more recent than the one of the former example, because its log key would be something like `wazuh-aws-wodle-configAWSLogs/XXXXXXXXXXXX/Config/us-east-1/2021/7/7/07/ConfigHistory...`,
and `11` < `7` when comparing strings.  

These are the reasons why the module is making one request per day specifying the exact month and date, beginning after the `only_logs_after` value. We think that this solution is too slow and should be improved since it makes requests for dates that have already been parsed, and we are currently working on designing a better solution.

**Comment by noise-kngdm** [comment]
## Issue update
We decided to modify the behavior of the module making one request per day, but also reducing the total amount of API requests to the minimum. While investigating how to implement this, we realized that the module was intended to work that way. The module would make one call to the `list_objects_v2` method every time the `iter_files_in_bucket` is executed, which will be executed once for every date returned by the `get_date_list` method, as the following code snippet states:

https://github.com/wazuh/wazuh/blob/175bd18f7cf58dd246a2407ef96eccafa0e83000/wodles/aws/aws_s3.py#L1555-L1557  

On the other hand, the `get_date_list` method works in the following way:

https://github.com/wazuh/wazuh/blob/175bd18f7cf58dd246a2407ef96eccafa0e83000/wodles/aws/aws_s3.py#L1096-L1100  

It creates a date list based on the last log's date. The reason why this method would make a request for every log beginning from the `only_logs_after` is that the query used to fetch the last log's date in the `get_date_last_log` was wrong:
https://github.com/wazuh/wazuh/blob/175bd18f7cf58dd246a2407ef96eccafa0e83000/wodles/aws/aws_s3.py#L422-L433  

It doesn't return the last log processed but **the first one** because the query results are being ordered in a descendant arrangement when they should be ordered in an ascendant one.  

This issue has been solved in #9993, so merging https://github.com/wazuh/wazuh/pull/10838 will solve this problem too. The other problem that affects the module's performance -which is that the module is making the same amount of API requests for unexistent account IDs-, is currently being tackled in #10840.
